src/Composer/Util/Zip.php

```diff
@@ -85,18 +86,15 @@ private static function locateFile(\ZipArchive $zip, $filename)
             if ($dirname === '.') {
                 $topLevelPaths[$name] = true;
                 if (\count($topLevelPaths) > 1) {
-                    // archive can only contain one top level directory
-                    return false;
+                    throw new \RuntimeException('Archive has more than one top level directories, and no composer.json was found on the top level, so it\'s an invalid archive. Top level path found were: '.implode(',', array_keys($topLevelPaths)));
                 }
-                continue;
             }
 
             // handle archives which do not have a TOC record for the directory itself
-            if (false === strpos('\\', $dirname) && false === strpos('/', $dirname)) {
+            if ($dirname !== '.' && false === strpos('\\', $dirname) && false === strpos('/', $dirname)) {
```

Comment:
This wasn't necessarily previously because of the `continue;` doing an "early return" out of the loop's block. I think it'd be preferable to restore the continue, or make this block an elseif, but duplicating the condition is suboptimal :)